### PR TITLE
feat: add a config prop 'aggressiveFollowUps' to follow up based on p…

### LINKED ISSUE #26: Priority Speed
Divide the config timeouts by the priority level. This makes follow ups more aggressive based on the priority level. 

`Priority: 5 (Emergency)` Means divide by 5. 

- warning = 3.5 days / 5 = 16.8 hours
-  disqualify = 7 days / 5 = 33.6 hours

Should be configurable (priority enabled: Boolean)

### PR DESCRIPTION
**Pull request by ishowvel**
Resolves #26

### CODE REVIEW

**Comment by ishowvel** [comment]
qa is in progress, please wait patiently

**Comment by ishowvel** [comment]
image
does anyone else get this error?
github is refusing to get my action released to marketplace
is there any other way to qa this?

**Comment by 0x4007** [comment]
You sure you need to publish it to use it?

**Comment by Keyrxng** [comment]
You shouldn't have to create a release this or even your kernel app.

To QA a plugin:

1. run the kernel locally
2. enter 
```yml
- plugins: 
  - uses:
    - plugin: http:localhost:4000
      with:
        ...plugginSettingsForThisplugin
```
3. run yarn worker inside the plugin repo if it's a worker-friendly plugin.
4. If it's an action plugin, you need to push your changes to the repo and instead of `localhost:4000` you use your repo name so `keyrxng/my-forked-plugin-repo:compute.yml@development` "compute.yml" is our workflow entry do not alter this part, "@dev" is whatever branch you are working on.
5. Trigger a github webhook event which `manifest.json` lists for the plugin and you should see the kernel catch the event in your terminal logs and then it passes it to `localhost:4000` directly to the plugin.

hope this helps

**Comment by ishowvel** [comment]
[development](https://github.com/ishowvel/daemon-disqualifier/actions/runs/11547443235/job/32137324733)
what may plugin github token be?
i couldnt find it in the readme nor in the compute yaml file

**Comment by gentlementlegen** [comment]
> [development](https://github.com/ishowvel/daemon-disqualifier/actions/runs/11547443235/job/32137324733) what may plugin github token be? i couldnt find it in the readme nor in the compute yaml file

@whilefoo RFC I think you can use `GITHUB_TOKEN` there.

**Comment by ishowvel** [comment]
> > [development](https://github.com/ishowvel/daemon-disqualifier/actions/runs/11547443235/job/32137324733) what may plugin github token be? i couldnt find it in the readme nor in the compute yaml file
> 
> @whilefoo RFC I think you can use `GITHUB_TOKEN` there.

It doesn't exist in the codebase as far as I know, it only exists in the polyfill in the dist dorectory

**Comment by ishowvel** [comment]
@gentlementlegen instead of qa's, can adding exhaustive tests be a good and acceptable alternative?

**Comment by gentlementlegen** [comment]
Both should be required. What is blocking you?

**Comment by ishowvel** [comment]
i have tried again and again to make this plugin work
heres the error this time 
```
TransformDecodeError: Invalid event [review_requested] (unknown event)
    at decode_Default (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:295144)
    at decode_FromArray (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:295232)
    at decode_Visit (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:297642)
    at decode_FromObject (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:296096)
    at decode_Visit (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:297782)
    at TransformDecode (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:298149)
    at Decode (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:305653)
    at createActionsPlugin (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:473956) {
  schema: {
    default: [
      'pull_request.review_requested',
      'pull_request.ready_for_review',
      'pull_request_review_comment.created',
      'issue_comment.created',
      'push'
    ],
    type: 'array',
    items: { type: 'string', [Symbol(TypeBox.Kind)]: 'String' },
    [Symbol(TypeBox.Kind)]: 'Array',
    [Symbol(TypeBox.Transform)]: { Decode: [Function (anonymous)], Encode: [Function (anonymous)] }
  },
  path: '/eventWhitelist',
  value: [ 'review_requested', 'ready_for_review', 'commented', 'committed' ],
  error: TypeBoxError: Invalid event [review_requested] (unknown event)
      at Object.Decode (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:570096)
      at decode_Default (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:295117)
      at decode_FromArray (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:295232)
      at decode_Visit (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:297642)
      at decode_FromObject (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:296096)
      at decode_Visit (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:297782)
      at TransformDecode (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:298149)
      at Decode (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:305653)
      at createActionsPlugin (file:///home/runner/work/daemon-disqualifier/daemon-disqualifier/dist/index.js:3:473956)
}
```
and heres the workflow [run](https://github.com/ishowvel/daemon-disqualifier/actions/runs/11570926907/job/32207714178)

**Comment by whilefoo** [comment]
This error originates from your Typebox Decode function. 

The values you supplied in the config are: `[ 'review_requested', 'ready_for_review', 'commented', 'committed' ]`,
but your Decode function is trying searching within `eventWhitelist` which has `[ "pull_request.review_requested", "pull_request.ready_for_review", "pull_request_review_comment.created", "issue_comment.created", "push"]` so it doesn't find it.
could you elaborate why you need to map these event names?

**Comment by ishowvel** [comment]
> This error originates from your Typebox Decode function. 
> 
> The values you supplied in the config are: `[ 'review_requested', 'ready_for_review', 'commented', 'committed' ]`,
> but your Decode function is trying searching within `eventWhitelist` which has `[ "pull_request.review_requested", "pull_request.ready_for_review", "pull_request_review_comment.created", "issue_comment.created", "push"]` so it doesn't find it.
> 
> Could you elaborate why you need to map these event names?
> 

I did figure that much, the thing is this isn't my piece of code. This piece of code is making it absolute hell for me

**Comment by Keyrxng** [comment]
@ishowvel You need to pass in the full webhook name `"pull_request.ready_for_review"` which then gets mapped to it's short name in relation to the timeline.

This [fn](https://github.com/ubiquity-os-marketplace/daemon-disqualifier/blob/9c6380912757547246e780f358431a8242ef6d13/src/types/plugin-input.ts#L39) takes values from this [object](https://github.com/ubiquity-os-marketplace/daemon-disqualifier/blob/9c6380912757547246e780f358431a8242ef6d13/src/types/plugin-input.ts#L27) which contain the full webhook name. This `map.get(webhook)` passes in the full webhook and the `Map` returns the short name.

[This](https://github.com/ubiquity-os-marketplace/daemon-disqualifier/blob/9c6380912757547246e780f358431a8242ef6d13/src/types/plugin-input.ts#L82) is where the schema is validated and rejected if incorrect.

hope this helps

**Comment by ishowvel** [comment]
Why are the short names in the valid configuration of this repository, shouldn't the valid configuration contain the full names of the events?

**Comment by Keyrxng** [comment]
@ishowvel Nice catch that's slipped through without being updated, yes you're right they should be updated I thought I remembered updating docs for that part, my mistake dude.

**Comment by ishowvel** [comment]
https://github.com/ishowvel/daemon-disqualifier/actions/runs/11602651701/job/32308069382
heres the qa, i added logging to show the normal warning and disqualification time vs the time it would take if prioritySpeed was enabled as you can see in the logs

**Comment by gentlementlegen** [comment]
@ishowvel Could you please remove the comma [here](https://github.com/ishowvel/daemon-disqualifier/actions/runs/11602651701/job/32308069382#step:4:23) so it doesn't break the URL?

File: src/helpers/task-metadata.ts [hunk 1]
```diff
@@ -108,3 +108,45 @@ function parseTimeLabel(
 
   return taskTimeEstimate;
 }
+
+export function parsePriorityLabel(
+  labels: (
```
Comment (gentlementlegen) [inline src/helpers/task-metadata.ts:113]: Pretty sure this type is defined within `octokit`.

File: src/helpers/task-metadata.ts [hunk 2]
```diff
@@ -108,3 +108,45 @@ function parseTimeLabel(
...
+    }
+
+    if (priorityLabel.startsWith("Priority:")) {
+      const matched = priorityLabel.match(/Priority: (\d+)/i);
+      if (!matched) {
+        return 0;
+      }
+
+      const [_, urgency] = matched;
+      taskPriorityEstimate = Number(urgency);
```
Comment (gentlementlegen) [inline src/helpers/task-metadata.ts:143]: Why not just returning here? Would avoid a `break`

File: src/helpers/task-update.ts [hunk 3]
```diff
@@ -75,13 +76,13 @@ export async function updateTaskReminder(context: ContextPlugin, repo: ListForOr
   if (lastReminderComment) {
     const lastReminderTime = DateTime.fromISO(lastReminderComment.created_at);
     mostRecentActivityDate = lastReminderTime > mostRecentActivityDate ? lastReminderTime : mostRecentActivityDate;
-    if (mostRecentActivityDate.plus({ milliseconds: disqualificationTimeDifference }) <= now) {
+    if (mostRecentActivityDate.plus({ milliseconds: prioritySpeed ? disqualificationTimeDifference / priorityLevel : disqualificationTimeDifference }) <= now) {
```
Comment (gentlementlegen) [inline src/helpers/task-update.ts:79]: This a a potential division by zero. Should be enforced to have positive values.

**Comment by ishowvel** [comment]
Done!

File: src/helpers/task-metadata.ts (hunk 1 above)
Comment (ishowvel) [reply src/helpers/task-metadata.ts]: This type wasn't defined exactly as is, I had to change one of the key value properties and make every key optional

File: src/helpers/task-metadata.ts [hunk 4]
```diff
@@ -2,6 +2,11 @@ import { DateTime } from "luxon";
 import ms from "ms";
 import { ListForOrg, ListIssueForRepo } from "../types/github-types";
 import { ContextPlugin } from "../types/plugin-input";
+import { RestEndpointMethodTypes } from "@octokit/rest";
+
+type IssueLabel = Partial<Omit<RestEndpointMethodTypes["issues"]["listLabelsForRepo"]["response"]["data"][0], "color">> & {
```
Comment (gentlementlegen) [inline src/helpers/task-metadata.ts:7]: Why is this needed? You seem to remove color to add it again.

File: src/helpers/task-update.ts [hunk 5]
```diff
@@ -46,6 +46,8 @@ export async function updateTaskReminder(context: ContextPlugin, repo: ListForOr
     .shift();
 
   const assignedDate = DateTime.fromISO(assignedEvent.created_at);
+  const priorityValue = parsePriorityLabel(issue.labels);
+  const priorityLevel = priorityValue === 0 ? priorityValue + 1 : priorityValue;
```
Comment (gentlementlegen) [inline src/helpers/task-update.ts:50]: You could use `Math.max` instead.

File: src/helpers/task-metadata.ts (hunk 4 above)
Comment (ishowvel) [reply src/helpers/task-metadata.ts:7]: The type from the endpoint isn't compatible with the labels being passed through as the type of color is strong | null in it and in this type fetched from endpoint, the type of color is strong | undefined

**Review by gentlementlegen — APPROVED** [review]
Please resolve the conflicts and good to go.

**Comment by ishowvel** [comment]
> Please resolve the conflicts and good to go.

Resolved!
